packages/runtime/container-runtime/src/test/garbageCollection.spec.ts

```diff
@@ -104,266 +68,583 @@ describe("Garbage Collection Tests", () => {
+        it("generates inactive events when inactive node is changed or revived", async () => {
+            const garbageCollector = createGarbageCollector();
+
+            // Remove node 2's reference from node 1. This should make node 2 and node 3 unreferenced.
+            defaultGCData.gcNodes[nodes[1]] = [];
+
+            await garbageCollector.collectGarbage({ runGC: true });
+
+            // Change all nodes.
+            changeAllNodes(garbageCollector);
+
+            // Validate that no inactive events are generated yet.
+            validateNoInactiveEvents();
+
+            // Wait for unreferenced timer (if any) to expire.
+            await waitForDeleteTimeout();
+
+            // Change all nodes. This should result in an inactiveObjectChanged event for node 2 and node 3 since they
+            // are inactive.
+            changeAllNodes(garbageCollector);
+            assert(
+                mockLogger.matchEvents([
+                    { eventName: inactiveObjectChangedEvent, deleteTimeoutMs, inactiveNodeId: nodes[2] },
+                    { eventName: inactiveObjectChangedEvent, deleteTimeoutMs, inactiveNodeId: nodes[3] },
+                ]),
+                "inactiveObjectChanged event not generated as expected",
+            );
+
+            // Add reference to node 3 from node 1.
+            defaultGCData.gcNodes[nodes[1]] = [ nodes[3] ];
+
+            // Run GC and validate that we get inactiveObjectRevived for node 3.
+            await garbageCollector.collectGarbage({ runGC: true });
+            assert(
+                mockLogger.matchEvents([
+                    { eventName: inactiveObjectRevivedEvent, deleteTimeoutMs, inactiveNodeId: nodes[3] },
+                ]),
+                "inactiveObjectRevived event not generated as expected",
+            );
+        });
 
-        // Remove node 2's reference from node 1. This should make node 2 and node 3 unreferenced.
-        defaultGCData.gcNodes[nodes[1]] = [];
+        it("generates inactive events once per node", async () => {
+            const garbageCollector = createGarbageCollector();
 
-        await garbageCollector.collectGarbage({ runGC: true });
+            // Remove node 3's reference from node 2.
+            defaultGCData.gcNodes[nodes[2]] = [];
 
-        // Change all nodes.
-        changeAllNodes(garbageCollector);
+            await garbageCollector.collectGarbage({ runGC: true });
 
-        // Validate that no inactive events are generated yet.
-        validateNoInactiveEvents();
+            await waitForDeleteTimeout();
 
-        // Wait for unreferenced timer (if any) to expire.
-        await waitForDeleteTimeout();
+            // Change all nodes. This should result in an inactiveObjectChanged event for node 2 and node 3 since they
+            // are inactive.
+            changeAllNodes(garbageCollector);
+            assert(
+                mockLogger.matchEvents([
```

Comment:
Something to consider logging structure can change over time. There are good and bad side-effects of using logging for testing. In some cases, it's better to have direct API / eventing system. It might be _public_ in a sense of particular layer (GC), but not exposed through other layers (like container runtime), thus keeping it mostly private / for internal FF usage.
I do not know if that's that case or not, just speaking out loudly :)